xarray/tests/test_coordinates.py

```diff
@@ -0,0 +1,141 @@
+from __future__ import annotations
+
+import pandas as pd
+import pytest
+
+from xarray.core.alignment import align
+from xarray.core.coordinates import Coordinates
+from xarray.core.dataarray import DataArray
+from xarray.core.dataset import Dataset
+from xarray.core.indexes import PandasIndex, PandasMultiIndex
+from xarray.tests import assert_identical, source_ndarray
+
+
+class TestCoordinates:
+    @pytest.fixture
+    def coords(self) -> Coordinates:
+        ds = Dataset(coords={"x": [0, 1, 2]})
+        return Coordinates(coords=ds.coords, indexes=ds.xindexes)
+
+    def test_init_noindex(self) -> None:
+        coords = Coordinates(coords={"foo": ("x", [0, 1, 2])})
+        expected = Dataset(coords={"foo": ("x", [0, 1, 2])})
+        assert_identical(coords.to_dataset(), expected)
+
+    def test_init_from_coords(self) -> None:
+        expected = Dataset(coords={"foo": ("x", [0, 1, 2])})
+        coords = Coordinates(coords=expected.coords)
+        assert_identical(coords.to_dataset(), expected)
+
+        # test variables copied
+        assert coords.variables["foo"] is not expected.variables["foo"]
+
+        # default index
+        expected = Dataset(coords={"x": ("x", [0, 1, 2])})
+        coords = Coordinates(coords=expected.coords, indexes=expected.xindexes)
+        assert_identical(coords.to_dataset(), expected)
+
+    def test_init_empty(self) -> None:
+        coords = Coordinates()
+        assert len(coords) == 0
+
+    def test_init_index_error(self) -> None:
+        idx = PandasIndex([1, 2, 3], "x")
+        with pytest.raises(ValueError, match="no coordinate variables found"):
+            Coordinates(indexes={"x": idx})
+
+        with pytest.raises(TypeError, match=".* is not an Xarray Index"):
+            Coordinates(coords={"x": ("x", [1, 2, 3])}, indexes={"x": "not_an_xarray_index"})  # type: ignore
+
+    def test_init_dim_sizes_conflict(self) -> None:
+        with pytest.raises(ValueError):
+            Coordinates(coords={"foo": ("x", [1, 2]), "bar": ("x", [1, 2, 3, 4])})
+
+    def test_from_pandas_multiindex(self) -> None:
+        midx = pd.MultiIndex.from_product([["a", "b"], [1, 2]], names=("one", "two"))
+        coords = Coordinates.from_pandas_multiindex(midx, "x")
+
+        assert isinstance(coords.xindexes["x"], PandasMultiIndex)
+        assert coords.xindexes["x"].index.equals(midx)
+        assert coords.xindexes["x"].dim == "x"
+
+        expected = PandasMultiIndex(midx, "x").create_variables()
+        assert list(coords.variables) == list(expected)
+        for name in ("x", "one", "two"):
+            assert_identical(expected[name], coords.variables[name])
+
+    def test_dims(self, coords) -> None:
```

Comment:
I would not use `coords` as a fixture on all these methods. Instead, I would re-create `coords` explicitly in each test.

The general guidance is only use generic fixtures/setup for creating objects if all somebody reading the code needs to know is that it's _some_ generic coordinates object, e.g., this would be OK:
```
def test_dataset_coords(self, coords):
  assert Dataset(coord).coords == coords
```

But here you're testing for particular values. For cases like this, it's better to make tests self-contained, even if they are a little redundant.